spec/index.adoc

```diff
@@ -443,6 +443,19 @@ If the canister has a https://webassembly.github.io/spec/core/binary/modules.htm
 +
 It is recommended for the canister to have a custom section called "icp:public candid:service", which contains the UTF-8 encoding of https://github.com/dfinity/candid/blob/master/spec/Candid.md#core-grammar[the Candid interface] for the canister.
 
+* `/canister/<canister_id>/history` (blob):
++
+The history of the canister. The value consists of a CBOR value encoding the canister history according to the following CDDL (the https://tools.ietf.org/html/rfc8610[Concise Data Definition Language (CDDL)] is a data description language for CBOR). You can also link:{attachmentsdir}/history.cddl[download the file].
```

Comment:
This puts all history into one leaf of the state tree.

Just wondering: Wouldn’t it be beneficial to put each change into the state tree individually, so once can query the latest only in a certified way?

In that case, I’d go for
```
/canister/<canister_id>/history/<change_index>/
```
where `<change_index>` is simply the (LEB-encoded) number of the index. 

But not sure if it’s important to support large histories efficiently.

Reply:
Could this not substantially increase the cost of certification? The history is trimmed to the point where it's feasible to return the entire history in a single response, so exposing all of it in one endpoint shouldn't add much complication to the caller,

Reply:
> Could this not substantially increase the cost of certification?

Not once [[MR-304]](https://dfinity.atlassian.net/browse/MR-304) is done.

Reply:
Works for me if performance won't be a problem. I guess we'll also need a `/canister/<canister_id>/history/latest/` that returns the latest change?

Reply:
I was thinking of `/canister/<canister_id>/history/total_num_changes` to give the total number of changes.

Reply:
> Works for me if performance won't be a problem.

Could we then store the entire history (and charge for it)?

Reply:
In the canisterOS meeting, we decided to store the individual change entries as separate paths in the system state tree. This has a number of benefits:
- in the future, we can easily increase the maximum number of entries kept in the history (we'll use 1000 in the implementation for now) without increasing the maximum size of a single value in the system state tree
- a dashboard caching canister history can periodically poll for the total number of changes and only fetch the missing change entries

while it is still easy to fetch the entire history by requesting the prefix `/canister/<canister_id>/history`.